### PR TITLE
Feature/gridedit 778 casulli derefinement

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/include/MeshKernel/CasulliDeRefinement.hpp [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+        /// @brief Clean up the edge
+        ///
+        /// @returns Indicates if the cleanp-up was successful or not
+        [[nodiscard]] bool CleanUpEdge(Mesh2D& mesh, const UInt edgeId) const;
+
+        /// @brief Find the id of the shared node for two edges.
+        ///
+        /// If no node is shared then return null value
+        UInt FindCommonNode(const Mesh2D& mesh, const UInt edgeId1, const UInt edgeId2) const;
```
Comment (lucacarniato) [inline libs/MeshKernel/include/MeshKernel/CasulliDeRefinement.hpp:163]: Is this function already implemented elsewhere?
meshkernel::UInt Mesh::FindCommonNode(UInt firstEdgeIndex, UInt secondEdgeIndex) const

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 2]
```diff
@@ -0,0 +1,1020 @@
...
+        UInt edgeId = mesh.m_facesEdges[elementId][i];
+
+        if (mesh.m_edgesNumFaces[edgeId] < 2)
+        {
+            continue;
+        }
+
+        UInt neighbouringElementId = elementId == mesh.m_edgesFaces[edgeId][0] ? mesh.m_edgesFaces[edgeId][1] : mesh.m_edgesFaces[edgeId][0];
+
+        if (std::ranges::find(connected, neighbouringElementId) == connected.end())
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:83]: do you need this check, std::ranges::find for every added face? it looks like you are simply looping over the edges of the faces. the added faces should not be the same if there are no hanging nodes

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 3]
```diff
@@ -0,0 +1,1020 @@
...
+
+    for (UInt i = 0; i < mesh.m_numFacesNodes[elementId]; ++i)
+    {
+        UInt nodeId = mesh.m_facesNodes[elementId][i];
+
+        for (UInt j = 0; j < mesh.m_nodesNumEdges[nodeId]; ++j)
+        {
+            UInt edgeId = mesh.m_nodesEdges[nodeId][j];
+
+            for (UInt k = 0; k < mesh.m_edgesNumFaces[edgeId]; ++k)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:105]: can you skip this loop and use a single if? m_edgesNumFaces can be at maximum 2

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 4]
```diff
@@ -0,0 +1,1020 @@
...
+        }
+
+        UInt elementId = mesh.m_edgesFaces[e][0];
+
+        if (mesh.m_numFacesNodes[elementId] != constants::geometric::numNodesInQuadrilateral)
+        {
+            continue;
+        }
+
+        if (ElementIsSeed(mesh, nodeTypes, elementId))
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:238]: maybe store the value of this function for each element?

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 5]
```diff
@@ -0,0 +1,1020 @@
...
+}
+
+std::vector<meshkernel::CasulliDeRefinement::ElementMask> meshkernel::CasulliDeRefinement::InitialiseElementMask(const Mesh2D& mesh,
+                                                                                                                 const std::vector<int>& nodeTypes) const
+{
+    UInt seedElement = FindElementSeedIndex(mesh, nodeTypes);
+    UInt iterationCount = 0;
+    UInt nMax = 1000;              // fix
+    UInt maxIterationCount = 1000; // fix
+
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:300]: nMax ,maxIterationCount as const initialized class members?

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 6]
```diff
@@ -0,0 +1,1020 @@
...
+    Point newNode(0.0, 0.0);
+    double factor = 0.0;
+
+    for (UInt i = 0; i < mesh.m_numFacesNodes[elementId]; ++i)
+    {
+        double fac = 1.0;
+        UInt nodeId = mesh.m_facesNodes[elementId][i];
+
+        if (nodeTypes[nodeId] == 2 || nodeTypes[nodeId] == 4)
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:519]: can we use an enum for nodeTypes? difficult to say what 2 and 4 are

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 7]
```diff
@@ -0,0 +1,1020 @@
...
+                        ++i;
+                    }
+
+                    if (k < mesh.m_numFacesNodes[connectedElementId])
+                    {
+                        std::shift_left(mesh.m_facesNodes[connectedElementId].begin() + i, mesh.m_facesNodes[connectedElementId].begin() + ndum, 1);
+                    }
+                    else
+                    {
+                        // Error: No node found
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:681]: do we need to throw an exception here?

File: libs/MeshKernel/src/CasulliDeRefinement.cpp [hunk 8]
```diff
@@ -0,0 +1,1020 @@
...
+            return false;
+        }
+    }
+
+    // deactivate cell
+    mesh.m_numFacesNodes[elementId] = 0;
+    return true;
+}
+
+meshkernel::UInt meshkernel::CasulliDeRefinement::FindCommonNode(const Mesh2D& mesh, const UInt edgeId1, const UInt edgeId2) const
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CasulliDeRefinement.cpp:915]: use what is already avail in mk

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/src/CasulliDeRefinement.cpp (hunk 8 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CasulliDeRefinement.cpp:915]: Removed

File: libs/MeshKernel/include/MeshKernel/CasulliDeRefinement.hpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+        /// @brief Get the most significant node type for all nodes of the element.
+        int GetNodeCode(const Mesh2D& mesh,
+                        const std::vector<int>& nodeTypes,
+                        const UInt elementId) const;
+
+        /// @brief Add element id to the list of id's
+        ///
+        /// only added is it is not already on the list and the element is a quadrilateral
+        void AddElementToList(const Mesh& mesh, const std::vector<UInt>& frontList, std::vector<UInt>& frontListCopy, const UInt kNew) const;
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CasulliDeRefinement.hpp:135]: Changed to elementId

File: libs/MeshKernel/include/MeshKernel/Mesh.hpp [hunk 10]
```diff
@@ -527,6 +542,16 @@ inline const meshkernel::Edge& meshkernel::Mesh::GetEdge(const UInt index) const
     return m_edges[index];
 }
 
+inline meshkernel::Edge& meshkernel::Mesh::GetEdge(const UInt index)
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/Mesh.hpp:545]: It is a little bit dangerous, because it bypasses the undo functionality

**Review by lucacarniato — APPROVED** [review]

File: libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp [hunk 11]
```diff
@@ -820,11 +820,41 @@ namespace meshkernelapi
         /// @returns Error code
         MKERNEL_API int mkernel_mesh2d_translate(int meshKernelId, double translationX, double translationY);
 
+        /// @brief Get list of elements that will be removed after the Casulli de-refinement algorithm
+        /// @param[in]     meshKernelId  The id of the mesh state
+        /// @param[in,out] elements      List of elements to be removed.
+        /// @returns Error code
+        MKERNEL_API int mkernel_mesh2d_casulli_derefinement_elements(int meshKernelId, GeometryList* elements);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp]: Removed the pointers.
Elements is passed back to the front end for plotting of the element centres

File: libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp [hunk 12]
```diff
@@ -820,11 +820,41 @@ namespace meshkernelapi
...
+        /// @param[in]     meshKernelId    The id of the mesh state
+        /// @param[in]     polygonGeometry The input polygon geometry
+        /// @param[in,out] elements        List of elements to be removed.
+        /// @returns Error code
+        MKERNEL_API int mkernel_mesh2d_casulli_derefinement_elements_on_polygon(int meshKernelId, GeometryList* polygonGeometry, GeometryList* elements);
+
+        /// @brief De-refine mesh using the Casulli de-refinement algorithm
+        /// @param[in]  meshKernelId  The id of the mesh state
+        /// @returns Error code
+        MKERNEL_API int mkernel_mesh2d_casulli_derefinement(int meshKernelId);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernelApi/include/MeshKernelApi/MeshKernel.hpp:886]: the element extension means that it computed only the element centres of the elements that are to be in the de-refinement. In interactor they are highlighted with a dot at the centre of the element

**Pull request merged**